src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,152 @@
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int addingLength = list.size();
+        T[] addingArrayList = (T[]) new Object[addingLength];
+        T[] updatedData = (T[]) new Object[size + addingLength];
+        resizeArray(data, 0, updatedData, 0, size);
+        for (int i = 0; i < addingLength; i++) {
+            addingArrayList[i] = list.get(i);
+        }
+        this.data = updatedData;
+        resizeArray(addingArrayList, 0, data, size, addingLength);
+        this.size += addingLength;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Impossible to get value to the pointed index"
+            );
+        }
+        return data[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Impossible to set value to the pointed index"
+            );
+        } else {
+            this.data[index] = value;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Impossible to remove value to the pointed index"
+            );
+        } else {
+            T removingElement = data[index];
+            removeElementByIndex(index);
+            return removingElement;
+        }
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int removingElementIndex = NOT_FOUND_ELEMENT;
+        for (int i = 0; i < data.length; i++) {
+            boolean isEqual = Objects.equals(element, data[i]);
+            if (isEqual) {
+                removingElementIndex = i;
+                break;
+            }
+        }
+        if (removingElementIndex == NOT_FOUND_ELEMENT) {
+            throw new NoSuchElementException("No element in the array list");
+        }
+        T removingElement = data[removingElementIndex];
+        removeElementByIndex(removingElementIndex);
+        return removingElement;
+    }
+
+    private void removeElementByIndex(int index) {
+        T[] updatedData = (T[]) new Object[data.length - 1];
+        resizeArray(data, 0, updatedData, 0, index);
+        resizeArray(data, index + 1, updatedData, index, data.length - index - 1);
+        this.data = updatedData;
+        size--;
+    }
+
+    private void resizeArray(
+            T[] arrayToAdd,
+            int arrayToAddStartIndex,
+            T[] takingArray,
+            int takingArrayStartIndex,
+            int elementsNumber
+    ) {
+        System.arraycopy(
+                arrayToAdd,
+                arrayToAddStartIndex,
+                takingArray,
+                takingArrayStartIndex,
+                elementsNumber
+        );
     }
```

Comment:
meaningless method. its not have any extra logic and have no differences between System.arraycopy
```suggestion
```